RadixWallet/Clients/ProfileStore/ProfileStore.swift

```diff
@@ -288,364 +223,305 @@ extension ProfileStore {
+	private static func _assertOwnership(
+		of profile: Profile,
+		against infoAboutThisDevice: DeviceInfo,
+		onMismatch: () -> Void
+	) throws {
+		loggerGlobal.debug("asserting ownership")
+		guard profile.header.lastUsedOnDevice.id == infoAboutThisDevice.id else {
+			let errorMessage = "Device ID mismatch, profile might have been used on another device. Last used in header was: \(String(describing: profile.header.lastUsedOnDevice)) and info of this device: \(String(describing: infoAboutThisDevice))"
+			loggerGlobal.error(.init(stringLiteral: errorMessage))
+			onMismatch()
+			throw Error.profileUsedOnAnotherDevice
 		}
+		// All good
 	}
 }
 
-// MARK: Internal
+// MARK: Private Static
 extension ProfileStore {
-	static func deviceDescription(
-		name: String,
-		model: DeviceFactorSource.Hint.Model
-	) -> NonEmptyString {
-		"\(name) (\(model.rawValue))"
-	}
-}
-
-// MARK: Private
-extension ProfileStore {
-	private static func restoreFromSecureStorageIfAble() async -> ProfileState {
-		@Dependency(\.jsonDecoder) var jsonDecoder
-		@Dependency(\.errorQueue) var errorQueue
-		@Dependency(\.secureStorageClient) var secureStorageClient
-		@Dependency(\.userDefaultsClient) var userDefaultsClient
-
-		let loadResult: Swift.Result<Profile?, Profile.LoadingFailure> = await {
-			guard let profileId = userDefaultsClient.getActiveProfileID() else {
-				return .success(nil)
-			}
-
-			guard
-				let profileSnapshotData = try? await secureStorageClient.loadProfileSnapshotData(profileId)
-			else {
-				return .success(nil)
-			}
-
-			let decodedHeader: ProfileSnapshot.Header
-			do {
-				// Implement decode version
-				decodedHeader = try ProfileSnapshot.Header.fromJSON(
-					data: profileSnapshotData,
-					jsonDecoder: jsonDecoder()
-				)
-			} catch {
-				return .failure(
-					.decodingFailure(
-						json: profileSnapshotData,
-						.known(.noProfileSnapshotVersionFoundInJSON
-						)
+	private static func _loadSavedElseNewProfile(
+		metaDeviceInfo: MetaDeviceInfo
+	) -> (deviceInfo: DeviceInfo, profile: Profile, conflictingOwners: ConflictingOwners?) {
+		let deviceInfo = metaDeviceInfo.deviceInfo
+		do {
+			if var existing = try _tryLoadSavedProfile() {
+				// Read: https://radixdlt.atlassian.net/l/cp/fmoH9KcN
+				let matchingIDs = existing.header.lastUsedOnDevice.id == deviceInfo.id
+				if metaDeviceInfo.fromDeprecatedDeviceID, matchingIDs {
+					// Same ID => migrate
+					existing.header.lastUsedOnDevice = deviceInfo
+				}
+				return (
+					deviceInfo: deviceInfo,
+					profile: existing,
+					conflictingOwners: matchingIDs ? nil : .init(
+						ownerOfCurrentProfile: existing.header.lastUsedOnDevice,
+						thisDevice: deviceInfo
 					)
 				)
+			} else {
+				return try (
+					deviceInfo: metaDeviceInfo.deviceInfo,
+					profile: _tryGenerateAndSaveNewProfile(deviceInfo: deviceInfo),
+					conflictingOwners: nil
+				)
 			}
+		} catch {
+			fatalError("Unable to use app. error: \(error)")
```

Comment:
should probably allow for an "escape hatch" here, I mentioned it on Slack, where user could see a list of keychain items with the label we use on mnemonics and be able to see their mnemonics.